internal/util/search/search.go

```
@@ -0,0 +1,178 @@
+		Filters: []kio.Filter{kio.FilterAll(sr)},
+		Outputs: []kio.Writer{inout},
+	}.Execute()
+}
+
+// Filter parses input node and performs search and replace operation on the node
+func (sr *SearchReplace) Filter(object *yaml.RNode) (*yaml.RNode, error) {
+	filePath, _, err := kioutil.GetFileAnnotations(object)
+	if err != nil {
+		return object, err
+	}
+	sr.filePath = filePath
+
+	if sr.shouldPutLiteralByPath() {
+		return object, sr.putLiteral(object)
+	}
+
+	// traverse the node to perform search/put operation
+	err = accept(sr, object)
+	return object, err
+}
+
+// visitMapping parses mapping node
+func (sr *SearchReplace) visitMapping(object *yaml.RNode, path string) error {
+	return nil
+}
+
+// visitSequence parses sequence node
+func (sr *SearchReplace) visitSequence(object *yaml.RNode, path string) error {
+	// TODO: pmarupaka support sequence nodes
+	return nil
+}
+
+// visitScalar parses scalar node
+func (sr *SearchReplace) visitScalar(object *yaml.RNode, path string) error {
+	regexMatch, err := sr.regexMatch(object.Document().Value)
+	if err != nil {
+		return err
+	}
+
+	pathMatch, err := sr.pathMatch(path)
+	if err != nil {
+		return err
+	}
+
+	valueMatch := object.Document().Value == sr.Value || regexMatch
+
+	// at least one of path or value must be matched
+	if (valueMatch && pathMatch) || (valueMatch && sr.Path == "") ||
+		(sr.Value == "" && sr.ValueRegex == "" && pathMatch) {
+		sr.Count++
+		if sr.PutLiteral != "" {
+			object.Document().Value = sr.PutLiteral
+		}
+		if sr.filePath != "" {
+			sr.Match[sr.filePath] = append(sr.Match[sr.filePath], fmt.Sprintf("%s: %s", strings.TrimPrefix(path, PathDelimiter), object.Document().Value))
+		}
+	}
+	return nil
+}
+
+// regexMatch checks if ValueRegex in SearchReplace struct matches with the input
+// value, returns error if any
+func (sr *SearchReplace) regexMatch(value string) (bool, error) {
+	if sr.ValueRegex == "" {
+		return false, nil
+	}
+	re, err := regexp.Compile(sr.ValueRegex)
```

You want to compile the regex once for invocation of the command. Currently, each node visited is calling compile which is a very expensive operation. In addition, if there's a error in compiling the regex, you want to fail fast instead of failing N times for each node.